linux/ibus-keyman/src/engine.c

```diff
@@ -692,16 +738,51 @@ process_capslock_action(
   return TRUE;
 }
 
-static gboolean process_end_action(IBusKeymanEngine *keyman) {
-  keyman->firstsurrogate = 0;
-  if (keyman->char_buffer != NULL) {
-    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
-    g_free(keyman->char_buffer);
-    keyman->char_buffer = NULL;
+static void
+commit_text(IBusKeymanEngine *keyman) {
+  g_assert(keyman != NULL);
+  if (keyman->commit_item > keyman->commit_queue) {
+    commit_queue_item *current_item = &keyman->commit_queue[0];
+    keyman->firstsurrogate = 0;
+    if (current_item->char_buffer != NULL) {
+      ibus_keyman_engine_commit_string(keyman, current_item->char_buffer);
+      g_free(current_item->char_buffer);
+    }
+    if (current_item->emitting_keystroke) {
+      ibus_engine_forward_key_event((IBusEngine*)keyman, current_item->keyval, current_item->keycode, current_item->state);
+    }
+    keyman->commit_item--;
+    memmove(keyman->commit_queue, &keyman->commit_queue[1], sizeof(commit_queue_item) * MAX_QUEUE_SIZE - 1);
+    initialize_queue(keyman, MAX_QUEUE_SIZE - 1, 1);
   }
-  if (keyman->emitting_keystroke) {
-    keyman->emitting_keystroke = FALSE;
-    return FALSE;
+}
+
+static gboolean
+process_end_action(IBusKeymanEngine *keyman) {
+  g_assert(keyman != NULL);
+  if (keyman->supports_prefilter) {
+    keyman->commit_item++;
+    if (keyman->commit_item > &keyman->commit_queue[MAX_QUEUE_SIZE-1]) {
+      g_error("Overflow of keyman commit_queue!");
+      // TODO: log to Sentry
+      keyman->commit_item--;
+    }
+
+    // Forward a fake key event to get the correct order of events so that any backspace key we
+    // generated will be processed before the character we're adding. We need to send a
+    // valid keyval/keycode combination so that it doesn't get swallowed by GTK but which
+    // isn't very likely used in real keyboards. F24 seems to work for that.
+    ibus_engine_forward_key_event((IBusEngine*)keyman, KEYMAN_NOCHAR_KEYSYM, KEYMAN_F24_KEYCODE_OUTPUT_SENTINEL, IBUS_PREFILTER_MASK);
```

Comment:
@mcdurdin Do we need to forward the fake key event if the app supports surrounding text?

Reply:
With the [prefilter version of ibus](https://github.com/keymanapp/ibus/pull/1) we get warnings with gedit:

```
(gedit:2437): Gdk-WARNING **: 17:52:19.388: Event with type 8 not holding a GdkDevice. It is most likely synthesized outside Gdk/GTK+

(gedit:2437): Gdk-WARNING **: 17:52:19.388: Event with type 9 not holding a GdkDevice. It is most likely synthesized outside Gdk/GTK+
```

Strangely enough we don't get these warnings with Chromium.

Reply:
Can we extend the comment above to reference the issue with Chromium doing its 'rate limiting' of hotkey-type keystrokes?